README.md
# @aminnairi/serve

Command-line interface for serving files.

[![Code Style](https://github.com/aminnairi/serve/workflows/Code%20Style/badge.svg)](https://github.com/aminnairi/serve/actions?query=workflow%3A%22Code+Style%22) [![Test](https://github.com/aminnairi/serve/workflows/Test/badge.svg)](https://github.com/aminnairi/serve/actions?query=workflow%3ATest) [![Package](https://github.com/aminnairi/serve/workflows/Package/badge.svg)](https://github.com/aminnairi/serve/actions?query=workflow%3APackage)

```console
$ npx @aminnairi/serve --folder client
Serving files from the client folder at http://127.0.0.1:8000. Hit CTRL+C at any time to stop.
```

## Requirements

- [Node 14.0.0 or later](https://nodejs.org/en/)
- [NPM](https://www.npmjs.com/)

## Usage

### NPX

```console
$ npx @aminnairi/serve --help
```

### NPM

```console
$ npm install @aminnairi/serve
$ node ./node_modules/.bin/serve --help
```

### Global

```console
$ sudo npm install --global @aminnairi/serve
$ serve --help
```

### Docker

```console
$ docker run --interactive --tty --user node --workdir /home/node --volume "$PWD":/home/node --publish 8000:8000 node npx @aminnairi/serve --host 0.0.0.0 --port 8000
```

### Docker Compose

```console
$ touch docker-compose.yaml
```

```yaml
version: "3"

services:
  server:
    image: node
    user: node
    working_dir: /home/node
    command: npx @aminnairi/serve --host 0.0.0.0 --port 8000
    ports:
      - 8000:8000
    volumes:
      - .:/home/node
```

```console
$ docker-compose up server
```

## Examples

### Folder

*The folder from where to serve the static folder (default to the current folder).*

```console
$ serve --folder public
```

### Host

*The host from which to listen (default to 127.0.0.1).*

```console
$ serve --host 0.0.0.0
```

### Port

*The port from which to listen (default to 8000). Fail if not an integer.*

```console
$ serve --port 5000
```

### Single-page application

*Whether to route all request to the current folder's index.html file (default to false).*

```console
$ serve --spa
```

### Base

*Set a base prefix url (default to empty) useful for GitHub Pages for instance.*

```console
$ serve --base aminnairi
```

### Verbose

*Display error message on the console instead of hidding them.*

```console
$ serve --verbose
```

### Help

*Display the help in the console.*

```console
$ serve --help
```


CHANGELOG.md
# Changelog

## Summary

- [`0.3.4`](#034)
- [`0.3.3`](#033)
- [`0.3.2`](#032)
- [`0.3.1`](#031)
- [`0.3.0`](#030)
- [`0.2.0`](#020)
- [`0.1.0`](#010)

## Versions

### 0.3.4

#### Breaking changes

None.

#### New features

None.

#### Minor changes, bug & security fixes

- Fixed trailing leading text functions ([#12](https://github.com/aminnairi/serve/pull/12))

### 0.3.3

#### Breaking changes

None.

#### New features

None.

#### Minor changes, bug & security fixes

- Fixed an issue with the log for when serving files from the current folder ([#10](https://github.com/aminnairi/serve/pull/10))

### 0.3.2

#### Breaking changes

None.

#### New features

None.

#### Minor changes, bug & security fixes

- Moved all dependencies to the dev dependencies ([#8](https://github.com/aminnairi/serve/pull/8))

### 0.3.1

#### Breaking changes

None.

#### New features

None.

#### Minor changes, bug & security fixes

- Fixed an issue with the name of the package in the documentation ([#6](https://github.com/aminnairi/serve/pull/6))

### 0.3.0

#### Breaking changes

None.

#### New features

-  Added a new base option ([#4](https://github.com/aminnairi/serve/pull/4))

#### Minor changes, bug & security fixes

None.

### 0.2.0

#### Breaking changes

None.

#### New features

- The dot folder is now accounted as the current folder ([#2](https://github.com/aminnairi/serve/pull/2))

#### Minor changes, bug & security fixes

None.

### 0.1.0

#### Breaking changes

None.

#### New features

None.

#### Minor changes, bug & security fixes

None.


test/serve.mjs
import chai from "chai";
import fetch from "node-fetch";
import {serve} from "../sources/server.mjs";
import {describe, it} from "mocha";

const {expect} = chai;

console.log = () => {
  // Mocked
};

console.error = () => {
  // Mocked
};

describe("server", () => {
  it("should open & close the server 127.0.0.1:8000", async () => {
    let server = null;

    try {
      server = await serve({base: "", folder: "sources", host: "127.0.0.1", port: 8000, spa: false, verbose: true});
      const response = await fetch("http://127.0.0.1:8000/mime-type.mjs");

      expect(response.ok).to.equal(true);
      server.close();
    } catch (error) {
      if (server) {
        server.close();
      }

      throw error;
    }
  });

  it("should open & close the server 127.0.0.1:5000", async () => {
    let server = null;

    try {
      server = await serve({base: "", folder: "sources", host: "127.0.0.1", port: 5000, spa: false, verbose: true});
      const response = await fetch("http://127.0.0.1:5000/mime-type.mjs");

      expect(response.ok).to.equal(true);
      server.close();
    } catch (error) {
      if (server) {
        server.close();
      }

      throw error;
    }
  });

  it("should open & close the server 0.0.0.0:5000", async () => {
    let server = null;

    try {
      server = await serve({base: "", folder: "sources", host: "0.0.0.0", port: 5000, spa: false, verbose: true});
      const response = await fetch("http://0.0.0.0:5000/mime-type.mjs");

      expect(response.ok).to.equal(true);
      server.close();
    } catch (error) {
      if (server) {
        server.close();
      }

      throw error;
    }
  });

  it("should fail when requesting another address other than the 127.0.0.1:8000", async () => {
    let server = null;

    try {
      server = await serve({base: "", folder: "sources", host: "127.0.0.1", port: 8000, spa: false, verbose: true});
      const response = await fetch("http://127.0.0.1:5000/mime-type.mjs");

      expect(response.ok).to.equal(true);
      server.close();
    } catch (error) {
      if (server) {
        server.close();
      }
    }
  });

  it("should return all urls to the index in spa mode", async () => {
    let server = null;

    try {
      server = await serve({base: "", folder: "example", host: "127.0.0.1", port: 8000, spa: true, verbose: true});
      const response1 = await fetch("http://127.0.0.1:8000/");
      const response2 = await fetch("http://127.0.0.1:8000/unknown");
      const response3 = await fetch("http://127.0.0.1:8000/hello-world");

      expect(response1.ok).to.equal(true);
      expect(response2.ok).to.equal(true);
      expect(response3.ok).to.equal(true);
      server.close();
    } catch (error) {
      if (server) {
        server.close();
      }

      throw error;
    }
  });
});


test/option.mjs
import chai from "chai";
import {getOptionValue} from "../sources/option.mjs";
import {describe, it} from "mocha";

const {expect} = chai;

describe("getOptionValue", () => {
  it("should get the correct parameter value", () => {
    const options = ["--folder", "public"];
    expect(getOptionValue({boolean: false, fallback: null, name: "--folder", options})).to.equal("public");
  });

  it("should return the fallback if the parameter value is not found", () => {
    const options = ["--folder"];
    expect(getOptionValue({boolean: false, fallback: null, name: "--folder", options})).to.equal(null);
  });

  it("should return the fallback if the parameter is not found", () => {
    const options = [];
    expect(getOptionValue({boolean: false, fallback: null, name: "--folder", options})).to.equal(null);
  });

  it("should return the fallback if the parameter is not found", () => {
    const options = [];
    expect(getOptionValue({boolean: false, fallback: null, name: "--folder", options})).to.equal(null);
  });

  it("should return the fallback if the parameter is not found", () => {
    const options = [];
    expect(getOptionValue({boolean: false, fallback: null, name: "--folder", options})).to.equal(null);
  });

  it("should return true if the parameter is a boolean", () => {
    const options = ["--verbose"];
    expect(getOptionValue({boolean: true, fallback: false, name: "--verbose", options})).to.equal(true);
  });

  it("should return false if the parameter is a boolean not found", () => {
    const options = [];
    expect(getOptionValue({boolean: true, fallback: false, name: "--verbose", options})).to.equal(false);
  });
});


test/string.mjs
import chai from "chai";
import {describe, it} from "mocha";
import {removeLeading, removeTrailing} from "../sources/string.mjs";

const {expect} = chai;

describe("string", () => {
  describe("removeLeading", () => {
    it("should remove the leading text", () => {
      const result = removeLeading("/", "/api");
      const expectation = "api";

      expect(result).to.equal(expectation);
    });
  });

  describe("removeTrailing", () => {
    it("should remove the trailing text", () => {
      const result = removeTrailing("/", "api/");
      const expectation = "api";

      expect(result).to.equal(expectation);
    });
  });
});


sources/server.mjs
import {createServer} from "http";
import {promises as fs} from "fs";
import {getMimeTypeOr} from "./mime-type.mjs";
import {join} from "path";
import {removeLeading, removeTrailing} from "./string.mjs";

const {readFile, lstat} = fs;

export const serve = ({folder, verbose, port, host, spa, base}) => {
  const separator = "/";
  const normalizedBase = removeTrailing(separator, removeLeading(separator, base));

  return new Promise(resolve => {
    const server = createServer(async (request, response) => {
      try {
        const requestPath = join(process.cwd(), folder, request.url.replace(normalizedBase, ""));
        const stat = await lstat(requestPath);
        const filePath = join(requestPath, stat.isDirectory() ? "index.html" : "");
        const fileBuffer = await readFile(filePath);
        const fileContent = fileBuffer.toString();

        return response.writeHead(200, {
          "Content-Type": getMimeTypeOr("text/plain", filePath)
        }).end(fileContent);
      } catch ({message}) {
        if (verbose) {
          console.error(message);
        }

        if (spa) {
          try {
            const filePath = join(process.cwd(), folder, "index.html");
            const fileBuffer = await readFile(filePath);
            const fileContent = fileBuffer.toString();

            return response.writeHead(200, {
              "Content-Type": getMimeTypeOr("text/plain", filePath)
            }).end(fileContent);
          } catch ({message: error}) {
            if (verbose) {
              console.error(error);
            }
          }
        }
      }

      return response.writeHead(404, {}).end();
    });

    server.listen(port, host, () => {
      const folderName = 0 === folder.length || "." === folder ? "current" : folder;
      console.log(`Serving files from the ${folderName} folder${base ? ` (with a base url of ${base})` : ""} at http://${host}:${port}. Hit CTRL+C at any time to stop.`);
      resolve(server);
    });

    server.on("error", ({message}) => {
      if (verbose) {
        console.error(message);
      }

      process.exit(1);
    });
  });
};


sources/serve.mjs
import {getOptionValue} from "./option.mjs";
import {serve} from "./server.mjs";

const options = process.argv.slice(2);

if (getOptionValue({boolean: true, fallback: false, name: "--help", options})) {
  console.log("--help: display this message.");
  console.log("--folder: The folder from where to serve the static folder (default to the current folder).");
  console.log("--host: The host from which to listen (default to 127.0.0.1).");
  console.log("--port: The port from which to listen (default to 8000). Fail if not integer.");
  console.log("--spa: Whether to route all request to the current folder's index.html file (default to false).");
  console.log("--verbose: Display error message on the console instead of hidding them.");
  console.log("--base: Set a base prefix url (default to empty) useful for GitHub Pages for instance.");
  console.log("Example: serve --folder public --host 0.0.0.0 --port 5000 --spa");

  process.exit(0);
}

const folder = getOptionValue({boolean: false, fallback: "", name: "--folder", options});
const host = getOptionValue({boolean: false, fallback: "127.0.0.1", name: "--host", options});
const port = Number(getOptionValue({boolean: false, fallback: "8000", name: "--port", options}));
const verbose = getOptionValue({boolean: true, fallback: false, name: "--verbose", options});
const spa = getOptionValue({boolean: true, fallback: false, name: "--spa", options});
const base = getOptionValue({boolean: false, fallback: "", name: "--base", options});

if (!Number.isInteger(port)) {
  console.error("The port must be an integer.");
  process.exit(1);
}

const main = async () => {
  const server = await serve({base, folder, host, port, spa, verbose});

  process.on("SIGINT", () => {
    server.close();
    console.log("Server stopped gracefully.");
    process.exit(0);
  });
};

main().catch(({message}) => {
  console.error(message);
});
